Deduplicate "edit bot" modal.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First four commits are prep commits to remove redundant code and some minor refactors and renaming.
- Next two commits are also refactor commits to modify the code for organization bot-edit form and moving it to `settings_bots.js`.
- Next two commits are to modify the organization bot-edit modal to allow changing avatar and config data of bots.
- Last two commits are to actually make both the modals same and remove the extra template file.

Fixes: <!-- Issue link, or clear description.--> #22033

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![bot-edit](https://user-images.githubusercontent.com/35494118/178332949-dbee8543-b739-48be-b30a-fa624180fab4.gif)

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

sahil839 commented on static/js/settings_bots.js:
```diff
@@ -510,20 +507,8 @@ export function set_up() {
-                                    .find("img")
-                                    .attr(
-                                        "src",
-                                        data.avatar_url + "&v=" + image_version.toString(),
```
Though we will be removing this code, but I added this commit in case anyone goes through commit history and finds that while reusing the code we do not take these lines into account and ignore them. Thus, I added this commit to correctly document that why we can remove this.

sahil839 commented on static/js/settings_bots.js:
```diff
@@ -282,6 +302,19 @@ export function set_up_bot_edit_form(bot) {
+                    $("#config_edit_inputbox input").each(function () {
+                        config_data[$(this).attr("name")] = $(this).val();
+                    });
+                    formData.append("config_data", JSON.stringify(config_data));
```
I am assuming we can allow admins to edit these configurations of bots not owned by them. Mentioning this since we did not discuss it before.

timabbott replied:
Hmm, the `config_data` could contain secrets (e.g. an API key for a third-party service), so we might want to go with the same approach that we have for API keys, which IIRC is that administrators need to take control of the bot explicitly by assigning themselves as the owner to access it? But maybe worth starting a #backend thread if I'm confused.

sahil839 commented on static/js/settings_bots.js:
```diff
@@ -230,7 +229,7 @@ export function set_up_bot_edit_form(bot) {
         full_name: bot.full_name,
         user_role_values: settings_config.user_role_values,
-        disable_role_dropdown: bot.is_owner && !page_params.is_owner,
+        disable_role_dropdown: !page_params.is_admin || (bot.is_owner && !page_params.is_owner),
```
Do we want to allow non-admins to change the role of bots owned by them? One way can be we allow them to only change it to a role lower than their role in the hierarchy, but I think that would be unnecessary complexity and currently role of bots do not matter much as in most of the cases we allow bots tasks that are allowed to their owners.

alya replied:
I think it's fine to only let admins change the role, as you've implemented.

timabbott replied:
Yeah, that sounds correct to me.

alya commented:
Also, I noticed that some of the vertical spacing between fields looks uneven:
<img width="308" alt="Screen Shot 2022-07-13 at 11 58 04 AM" src="https://user-images.githubusercontent.com/2090066/178810630-b27364c1-4b25-40b7-b475-9098f29275a1.png">

alya commented:
Otherwise looks good to me!

timabbott commented:
I merged the first few commits as the series ending with 0d20d132b7df4c51f266acfaf086f8d7b7b2420b.

ganpa3 commented:
Clicking on the "Full Name" and the "Avatar" label closes the modal.

![avatar_closes_modal](https://user-images.githubusercontent.com/58626718/179154802-fda9bede-9915-422a-bbcd-f8ad46dfa532.gif)

ganpa3 commented:
In the personal edit bot form, the dropdown overflows the modal, and the modal enlarges causing the buttons to "shift" up. This happens on small screens like laptop

![image](https://user-images.githubusercontent.com/58626718/179155295-dabd108e-77c0-4b73-b119-fc428571d637.png)

ganpa3 commented on static/js/settings_bots.js:
```diff
@@ -220,6 +223,130 @@ export function update_bot_permissions_ui() {
+            success() {
+                avatar_widget.clear();
+                dialog_widget.close_modal();
+                dialog_widget.hide_dialog_spinner();
```
small nit: `close_modal()` removes the modal from the DOM, so calling `hide_dialog_spinner` after is redundant.

sahil839 replied:
Ah yes, fixed this.

sahil839 commented:
Thanks for the review everyone.
- Fixed the spacing and fonts.
- Fixed the problem of modal closing when clicking on labels that was due to incorrect "for" attribute on them. I am not sure whether we need "for" attribute on the avatar label, If we set it correctly to id of the button, then the window for selecting picture opens on clicking the label. Can add the attribute, if we want this behavior. @timabbott Also one thing that currently defining "for" attribute is not consistent throughout the app, we should try to do that. In many places we skip it and in some places we set its value to a non-existent ID.
- Regarding the issue for the owner dropdown overflowing, I think this is related to #22169. And as mentioned about enlarging of modal, I cannot reproduce it if I first make the screen shorter and then open the dropdown. Though I can reproduce it If I keep the dropdown on full sized window in my laptop and then decrease the height of browser window. And clipping of dropdown can be reproduced at all sizes. And both of the above irrespective of modal in personal settings and admin settings. @ganpa3 can you confirm this?

sahil839 commented:
Rebased the PR. Changes are -
- Now deactivate button is added to the modal opened from personal settings page as well. Though we already have deactivate option in the panel, but it is same in other cases as well where we have deactivate option in main panel and in edit modal too.
- Also added confirmation modal when deactivating from the icon on the bot card in "#settings/bots" panel.

alya commented:
Re-reviewed this PR -- looks good to me! Should be ready for @timabbott once conflicts are resolved.

sahil839 commented:
Fixed the tests, ready for review by Tim.

timabbott commented:
This had an `input_group` inside another `input_group` when using the `service_data` setup, which seemed not intended. Adding this hunk to the last commit looks identical to what you had, so I applied it:

```
diff --git a/static/templates/settings/edit_bot_form.hbs b/static/templates/settings/edit_bot_form.hbs
index d6ea0bd0b2..19c4cf3246 100644
--- a/static/templates/settings/edit_bot_form.hbs
+++ b/static/templates/settings/edit_bot_form.hbs
@@ -27,7 +27,7 @@
               widget_name="edit_bot_owner"
               list_placeholder=(t 'Filter users')}}
         </div>
-        <div id="service_data" class="input-group">
+        <div id="service_data">
         </div>
         <div class="input-group edit-avatar-section">
             <label>Avatar</label>
```

Merged with that tweak, huge thanks for doing this @sahil839! It is very satisfying to remove that much surface messy area.
